earth2studio/statistics/rank.py

```diff
@@ -0,0 +1,157 @@
+    number_of_bins: int
+        The number of bins to discretize the unit interval over.
+        by default, 10
+    """
+
+    def __init__(
+        self,
+        ensemble_dimension: str,
+        reduction_dimensions: list[str],
+        number_of_bins: int = 10,
+    ):
+        if not isinstance(ensemble_dimension, str):
+            raise ValueError(
+                "Error! Rank Histogram currently assumes reduction over a single dimension."
+            )
+
+        self.ensemble_dimension = ensemble_dimension
+        self._reduction_dimensions = reduction_dimensions
+        self.number_of_bins = number_of_bins
+
+    def __str__(self) -> str:
+        return "rank_histogram"
+
+    @property
+    def reduction_dimensions(self) -> list[str]:
+        return [self.ensemble_dimension] + self._reduction_dimensions
+
+    def __call__(
+        self,
+        x: torch.Tensor,
+        x_coords: CoordSystem,
+        y: torch.Tensor,
+        y_coords: CoordSystem,
+    ) -> tuple[torch.Tensor, CoordSystem]:
+        """
+        Apply metric to data `x` and `y`, checking that their coordinates
+        are broadcastable. While reducing over `reduction_dims`.
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            Input tensor #1 intended to apply metric to.
+        x_coords : CoordSystem
+            Ordered dict representing coordinate system that describes the `x` tensor.
+            'reduction_dims' must be in coords.
+        y : torch.Tensor
+            Input tensor #2 intended to apply statistic to.
+        y_coords : CoordSystem
+            Ordered dict representing coordinate system that describes the `y` tensor.
+            'reduction_dims' must be in coords.
+
+        Returns
+        -------
+        tuple[torch.Tensor, CoordSystem]
+            Returns rank histogram tensor with appropriate reduced coordinates.
+        """
+        if not all([rd in x_coords for rd in self.reduction_dimensions]):
+            raise ValueError(
+                "Initialized reduction dimension does not appear in passed coords"
+            )
+
+        # Do some coordinate checking
+        # Assume ensemble_dim is in x_coords but not y_coords
+        if self.ensemble_dimension in y_coords:
+            raise ValueError(
+                f"{self.ensemble_dimension} should not be in y_coords but is."
+            )
+        if x.ndim != y.ndim + 1:
+            raise ValueError(
+                "x and y must have broadcastable shapes but got"
+                + f"{x.shape} and {y.shape}"
+            )
+        # Input coordinate checking
+        coord_count = 0
+        for c in x_coords:
+            if c != self.ensemble_dimension:
+                handshake_dim(y_coords, c, coord_count)
+                handshake_coords(y_coords, x_coords, c)
+                coord_count += 1
+
+        dim = list(x_coords).index(self.ensemble_dimension)
+
+        _ranks = torch.mean(1.0 * torch.ge(y, x), dim=dim)
```

Comment:
Bit confused by this here.

From my understanding, the ensemble elements should define the bin edges. The rank seems to be the average of the reduction dims not ensemble?

https://journals.ametsoc.org/view/journals/mwre/129/3/1520-0493_2001_129_0550_iorhfv_2.0.co_2.xml

Reply:
In that paper they say, "A rank histogram is found by repeatedly tallying the rank of the truth relative to an actual distribution of sorted ensemble forecasts (which may or may not be calibrated)". Then we average over a large sample of statistically independent points.

So first, at a given point, we find the rank (between 0 and 1) of the observation with respect to the ensemble. Now we have reduced the ensemble dimension through the rank procedure (this is what occurs on line 125). Then, we take a histogram over all of the ranks over some set of statistically independent points (this is the reduction dimension).